### PR TITLE
padded dataloader

### PR DESCRIPTION
**Pull request by lapp0**
As discussed in https://github.com/Synthyra/SpeedRunningESM2/pull/3

This removes the issue of a smaller batch size resulting in a larger number of truncations. Should be used for validation / test. Unclear whether it should be used for training - it's more sample efficient, but 4% slower and has fewer tokens per batch.

### CODE REVIEW

File: dataloading.py [hunk 1]
```diff
@@ -56,3 +56,34 @@ def next_batch(self):
...
         return input_ids
+
+
+class DistributedPaddedDataLoader(DistributedDataLoader):
+    def __init__(self, filename_pattern, seq_len, process_rank, num_processes, eos_id, pad_id):
+        super().__init__(filename_pattern, seq_len, process_rank, num_processes)
+        assert (eos_id is None) == (pad_id is None)
+        self.eos_id = eos_id
+        self.pad_id = pad_id
+        self.padding = (eos_id is not None) and (pad_id is not None)
```
Comment (lapp0) [inline dataloading.py:67]: should remove

File: train_esm2.py [hunk 2]
```diff
@@ -161,7 +161,8 @@ def print0(s, logonly=False):
     # load tokens
     train_loader = DistributedDataLoader(args.input_bin, batch_size, ddp_rank, ddp_world_size)
     valid_loader = DistributedDataLoader(args.input_valid_bin, batch_size, ddp_rank, ddp_world_size)
-    test_loader = DistributedDataLoader(args.input_test_bin, batch_size, ddp_rank, ddp_world_size)
+    test_loader = DistributedDataLoader(args.input_test_bin, batch_size // 8, ddp_rank, ddp_world_size)
+    test_loader_padded = DistributedPaddedDataLoader(args.input_test_bin, batch_size // 8, ddp_rank, ddp_world_size, eos_id=2, pad_id=1)
```
Comment (lapp0) [inline train_esm2.py:165]: shouldn't hardcode

File: train_esm2.py [hunk 3]
```diff
@@ -331,14 +332,26 @@ def get_lr(it):
...
+    test_loader_padded.reset()
+    test_loss_padded = 0.0
+    with torch.no_grad():
+        for _ in range(test_steps):
+            input_ids = test_loader_padded.next_batch()
+            test_loss_padded += model(input_ids, sliding_window_size)
+    test_loss_padded /= test_steps
+    print0(f"Test (Padded) results | Loss: {test_loss_padded:.4f} | Perplexity: {math.e**test_loss_padded:.4f}")
+    #############
+
```
Comment (lapp0) [inline train_esm2.py:354]: remove, for demonstration only

File: dataloading.py [hunk 4]
```diff
@@ -56,3 +56,32 @@ def next_batch(self):
...
+        self.current_shard = (self.current_shard + self.num_processes) % len(self.files)
+        self.current_position = 0
+        self.tokens = _load_data_shard(self.files[self.current_shard], self.files_num_tokens[self.current_shard])
+
+    def next_batch(self):
+        end_pos = self.current_position + self.batch_size
+        buf = self.tokens[self.current_position:end_pos]
+        input_ids = buf.to(device="cuda", dtype=torch.int32, non_blocking=True)
+        keep = (input_ids == self.eos_id).cumsum(dim=0).argmax().item()
+        keep = max(keep or 0, self.batch_size - 2048)
```
Comment (lapp0) [inline dataloading.py:81]: max sliding window size - should be a parameter

**Comment by lapp0** [comment]
This code has a few bugs. I think it makes more sense to simply update the data packing script so each `.bin` file ends with EOS, and begins with CLS.

Does that make sense to you?

**Comment by lhallee** [comment]
> This code has a few bugs. I think it makes more sense to simply update the data packing script so each `.bin` file ends with EOS, and begins with CLS.
> 
> Does that make sense to you?

Yep, just pushed this. Processing the new packed dataset now.

Do you think we should add the padding just for the eval or for training too? Seems like in training we could still use `keep` to keep track of what full sequences have been input but just don't add the pad tokens (more real tokens better?).

**Comment by lapp0** [comment]
I've seen better sample efficiency with padded training, although it was with a low sample size and needs further verification.

I'm thinking we should avoid padding within the packed dataset itself. Correct padding behavior is tied to batch size - if the batch size is smaller than the  batch size used for packed dataset padding, we will be truncating. If the batch size is larger, we end up having padding within a sample, which breaks `Rotary` behavior.

I'm pushing changes to construct the padding at runtime which overrides the `advance()` method. Additionally it fixes a bug where we train on the last sequence in the test set due to overwritten `input_ids` var.

You previously stated there were 2,077,660 validation tokens, however when attempting to verify the integrity I see 2,073,745:

```
step:0/20000 val_loss:3.4965 train_time:0ms step_avg:nanms perplexity:33.0000 param_count:43,776,054 tokens: 2073745
```

I see that `Synthyra/omg_prot50_packed` has been updated since. Can you check whether this number is still accurate?

File: dataloading.py [hunk 5]
```diff
@@ -2,57 +2,120 @@
...
+        seq = super().next_batch()
+
+        # TODO: for verification of proper padding, can remove
+        if False:
+            # first token has to be BOS or there is no BOS
+            assert seq[0] == 0 or (seq == 0).sum() == 0, "first bos or no bos"
+            # last token has to be EOS or there is no EOS
+            first_pad_idx = (seq == self.pad_id).nonzero(as_tuple=True)[0][0].item()
+            assert (seq == self.eos_id).sum() == 0 or seq[first_pad_idx-1] == self.eos_id
+
```
Comment (lapp0) [inline dataloading.py:120]: Can remove - demonstrates the verification steps taken for loaded sequences.

File: train_esm2.py [hunk 6]
```diff
@@ -249,15 +246,19 @@ def get_lr(it):
             model.eval()
             valid_loader.reset()
             val_loss = 0.0
+            valid_steps = 0
+            valid_tokens = 0
             with torch.no_grad():
-                for _ in range(valid_steps):
-                    input_ids = valid_loader.next_batch()
```
Comment (lapp0) [inline train_esm2.py:254]: overwriting `input_ids` results in training on the test set every val step.

Comment (lhallee) [reply train_esm2.py:254]: We can just move `input_ids = train_loader.next_batch()` to right before we pass input ides to model during training right?

**Comment by lhallee** [comment]
@lapp0 the new shards are in the process of uploading, some error made them fail yesterday.

Overall the code looks good, I think I'll start implementing the metrics again now that we can use reduced batch size and get accurate numbers.

**Comment by lapp0** [comment]
Thanks!

A new packed data isn't necessary. This code assumes no padding and works with the current packed dataset.

A few more fixes:
- I figured out the issue with the token counts referenced in my previous comment (I was missing the very first sample). Will be pushing a separate commit fixing this bug.
- Weighing loss based on non-padded token count. Although it's worth mentioning that in the training set the gradients aren't weighted by their non-pad token count.
- Don't add 1 to self.local_batch_size, this is an artifact of modded-nanogpt since it's loss objective involves shifting by one token between input and target.

Some tokens are still missing, this is an artifact of modded-nanogpt's behavior. It ignores the last `file_token_count % batch_size` tokens. This should be fine.

Also worth mentioning the padding ratio for the val set:

2,077,660 tokens before padding, 2,162,688 after.
